### DISCUSSION TITLE
Resizing dynamic centered column

### DISCUSSION THREAD

**Discussion by yngwi** [fltk-rs/fltk-rs#1306] [Q&A] [answered]
Hi,
I'm very new to fltk-rs and I'm trying to create a layout that has a centered scrolling column that has a limited with. The central column should stay centered even when resizing the window. I have added something like this:

```rust
use fltk::{app, enums::FrameType, frame, group, prelude::*, window};

fn main() {
    let a = app::App::default();
    let mut win = window::Window::default().with_size(400, 300);
    let scroll = group::Scroll::default_fill().with_type(group::ScrollType::Vertical);
    let mut content = group::Pack::default_fill().with_type(group::PackType::Vertical);
    content.set_spacing(10);
    for index in 0..100 {
        let mut container = group::Flex::default()
            .with_size(0, 30)
            .with_type(group::FlexType::Row);
        frame::Frame::default();
        let mut text = frame::Frame::default().with_label(&index.to_string()[..]);
        text.set_frame(FrameType::BorderBox);
        frame::Frame::default();
        container.set_size(&text, 200);
        container.end();
    }
    content.end();
    scroll.end();
    win.make_resizable(true);
    win.resize_callback(move |_win, _x, _y, _w, _h| {
        // Do what?
    });
    win.end();
    win.show();
    a.run().unwrap();
}
```

When launching, it looks like this:
![image](https://user-images.githubusercontent.com/963440/200129524-48944d01-72a0-495e-9245-f5bd8cb1aa97.png)

After resizing like this:
![image](https://user-images.githubusercontent.com/963440/200129540-99fc13b8-d5eb-4e28-885b-bf7a6d8dc121.png)

I assume I have to enter the resizing code into the resize_callback for the window, but I couldn't find out so far what to add to still keep the layout dynamic. I'm grateful for any input or pointers to an example.

Best,
Daniel

**Comment by MoAlyousef** [comment, marked as answer]
Hi,
You can try
```rust
    win.resize_callback(move |_win, _x, _y, w, h| {
        content.set_size(w, h);
    });
```

The idea is several group type widgets (like Group and Scroll) intentionally ignore parent resizing when it comes to their children. It allows creating custom widgets and such.
You were right in reaching for resize_callback, whose captures are: the widget being resized, it's x, y, w, h.

**Reply by yngwi** [reply]
So, so I have to calculate the new size manually or do other widgets also fire it when the main window is resized?

**Reply by MoAlyousef** [reply]
The resize event is triggered by the window, so it's best to use the window's resize_callback, you can however use other any widget's resize_callback. The w, h in the example is automatically determined by FLTK, so you wouldn't need to calculate things, except if you were overriding FLTK's relative positioning.

**Reply by yngwi** [reply]
Thank you very much! I tried the code snippet now, it works very well if the scroll position is at the top but if the scroll bar is down a little the contant vanishes and the scroll-handle size changes, like this:

![image](https://user-images.githubusercontent.com/963440/200139151-f63e4b93-2f85-431e-8f23-33f841df0cf6.png)

After scrolling again, the content goes back to normal. Could you tell me the reason for this and if there is a way to avoid this or should I reset the scroll position before doing the resizing in the resize_callback? Again, thank you for your help!

**Reply by MoAlyousef** [reply]
hmm I'm seeing this and I'm not sure what might be the cause. I'll try to ask in the FLTK mailing list. Meanwhile, scrolling to 0 seems to prevent this issue:
`scroll.scroll_to(0, 0);`

**Reply by MoAlyousef** [reply]
Also for some reason if you use resize it will automatically scroll to 0:
```rust
    win.resize_callback(move |_win, _x, _y, w, h| {
        content.resize(0, 0, w, h);
    });
```

**Reply by yngwi** [reply]
Thank you!

**Reply by yngwi** [reply]
@MoAlyousef Hi, I have another related question: I'm trying to use the resize callback to also send a message about the window being resized to then store the last used position / size on the disk so I can remember this for the next restart. Would this be the right way to do this? Firing an event (as described in the events section in the book) does work for other events like a button press but not for this resize event. Strangly the event seems to be fired only when I later click on the close window control.

**Reply by MoAlyousef** [reply]
I would probably do this in the window’s set_callback method. This is triggered when you press the close button, so you would also have to handle that:
```rust
win.set_callback(|win| {
   // save win.w() and win.h() to some file
   app.quit();
});
```

**Reply by yngwi** [reply]
That makes sense, thank you!
